### PR TITLE
Update README

### LINKED ISSUE #154: Update README to indicate controller architecture changes
1. Common controller will watch VolumeSnapshotContent and bind VolumeSnapshotContent with VolumeSnapshot.
2. Snapshotter will watch VolumeSnapshot to invoke corresponding gRPC call, create VolumeSnapshotContent.
3. Deletion of a snapshot should follow similar pattern

### PR DESCRIPTION
**Pull request by xing-yang**
**What type of PR is this?**
/kind documentation

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #154

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,12 +2,19 @@
...
 
 ## Overview
 
 CSI Snapshotter is an external controller that watches Kubernetes Snapshot CRD objects and triggers CreateSnapshot/DeleteSnapshot against a CSI endpoint. Full design can be found at Kubernetes proposal at [here](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-snapshot.md)
 
+With the promotion of Volume Snapshot to beta, the feature is now enabled by default on standard Kubernetes deployments instead of being opt-in.
+
+The move of the Kubernetes Volume Snapshot feature to beta also means:
+* A revamp of volume snapshot APIs.
+* The CSI external-snapshotter sidecar is split into two controllers, a common snapshot controller and a CSI external-snapshotter sidecar.
```
Comment (msau42) [inline README.md]: I would remove the word "common" and just call it snapshot controller

File: README.md [hunk 2]
```diff
@@ -30,23 +37,30 @@ Snapshotter talks to CSI over socket (/run/csi/socket by default, configurable b
   * If the `VolumeSnapshotRef` is set to nil, skip this content since it is not bound to any snapshot object.
   * Otherwise, the controller verifies whether the content object is correctly bound to a snapshot object. In case the `VolumeSnapshotRef.UID` is set but it does not match its snapshot object or snapshot no long exists, the content object and its associated snapshot will be deleted.
 
-## Usage
+## Deployment
```
Comment (msau42) [inline README.md]: Do we also want to provide some recommended RBACs for users (not sidecar, not controller) to operate on the CRs?

File: README.md (hunk 1 above)
Comment (xing-yang) [reply README.md]: Removed

File: README.md (hunk 2 above)
Comment (xing-yang) [reply README.md]: Does that include admins?  Users can only see VolumeSnapshot, Admins can see both.

Comment (msau42) [reply README.md]: Correct, I mean provide different recommended rbacs for different personas

File: README.md [hunk 3]
```diff
@@ -1,52 +1,75 @@
...
+The snapshot controller is deployed by the Kubernetes distributions and is responsible for watching the VolumeSnapshot CRD objects.
 
-### Snapshotter
+The CSI external-snapshotter sidecar watches Kubernetes VolumeSnapshotContent CRD objects and triggers CreateSnapshot/DeleteSnapshot against a CSI endpoint.
 
-Snapshotter talks to CSI over socket (/run/csi/socket by default, configurable by -csi-address). The snapshotter then:
+Blog post for the beta feature can be found [here](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-cis-volume-snapshot-beta)
 
-* Discovers the supported snapshotter name by `GetDriverName` call.
+## Controller Workflow
```
Comment (msau42) [inline README.md]: How much of this is relevant to a vendor looking to integrate these components, and how much can go into a "design.md" (or even KEP) file?

File: README.md [hunk 4]
```diff
@@ -55,12 +78,24 @@ If you want to run external-snapshotter with higher availability, you can enable
 --leader-election=true
 ```
 
+### Upgrade from Alpha to Beta
+
+The change from Alpha to Beta snapshot APIs is not backward compatible.
+
+If you have already deployed Alpha snapshot APIs and controller and want to upgrade to Beta, you need to do the following:
```
Comment (msau42) [inline README.md]: You can also set the deletepolicy to retain, delete the alpha objects, and then import them as pre-existing snapshots using the beta APIs

File: README.md [hunk 5]
```diff
@@ -1,52 +1,75 @@
-# CSI Snapshotter
+# CSI Snapshot Feature
```
Comment (msau42) [inline README.md]: Can you also mirror the README format that the other sidecars have?

File: README.md [hunk 6]
```diff
@@ -1,52 +1,75 @@
...
 
-The CSI external-snapshotter is part of Kubernetes implementation of [Container Storage Interface (CSI)](https://github.com/container-storage-interface/spec).
+The CSI snapshot feature is part of Kubernetes implementation of [Container Storage Interface (CSI)](https://github.com/container-storage-interface/spec).
```
Comment (msau42) [inline README.md]: I see a few main audiences for this readme:

1. Users wanting to use the snapshot feature.
2. CSI driver authors wanting to integrate snapshot support for their driver
3. Kubernetes distributions/deployers that want to support snapshots.
4. Developers making changes to the components.

For 1), I would point them to the K8s docs.

And then clearly state the audience for the rest of this README, which I think should be 2) and 3).

For 4), we should have another doc like "DEVELOPING.md"

File: README.md (hunk 3 above)
Comment (xing-yang) [reply README.md]: Most of this section is actually already in the blog post.  I added it here because in the old README, there's some design overview on how the controller is implemented.  I can just remove this and refer to the blog which is already linked.  I can also submit a KEP later if that's preferred.

File: README.md (hunk 6 above)
Comment (xing-yang) [reply README.md]: Sure.  I'll take a look of README of other sidecars.  I think the old README mirrored the one in external-attacher.

File: README.md (hunk 3 above)
Comment (msau42) [reply README.md]: I think we should capture detailed design and workflows somewhere other than a blog post. I think KEP is a natural place for this

File: README.md (hunk 5 above)
Comment (xing-yang) [reply README.md]: With Beta, I'll have to talk about the snapshot controller and the CSI snapshotter.  So it won't be exactly the same as other sidecars.

File: README.md [hunk 7]
```diff
@@ -1,66 +1,145 @@
...
-External snapshotter follows [controller](https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md) pattern and uses informers to watch for `VolumeSnapshot` and `VolumeSnapshotContent` create/update/delete events. It filters out these objects with `Snapshotter==<CSI driver name>` specified in the associated VolumeSnapshotClass object and then processes these events in workqueues with exponential backoff.
 
-### Snapshotter
+## Compatibility
 
-Snapshotter talks to CSI over socket (/run/csi/socket by default, configurable by -csi-address). The snapshotter then:
+This information reflects the head of this branch.
 
-* Discovers the supported snapshotter name by `GetDriverName` call.
+| Compatible with CSI Version                                                                | Container Image             | Min K8s Version |
```
Comment (msau42) [inline README.md]: Do we also need a column for what versions of the snapshot CRDs it supports?

File: README.md [hunk 8]
```diff
@@ -1,66 +1,145 @@
...
-for i in $(find deploy/kubernetes -name '*.yaml'); do kubectl create -f $i; done
-```
+Install CSI Driver:
+* Follow instructions provided by your CSI Driver vendor.
+* Here is an example to install the sample hostpath CSI driver
+  * kubectl create -f deploy/kubernetes/csi-snapshotter
+  * https://github.com/kubernetes-csi/external-snapshotter/tree/master/deploy/kubernetes/csi-snapshotter
 
-### Running with Leader Election
+### Important changes in the snapshot beta APIs
```
Comment (msau42) [inline README.md]: clarify from alpha1v1-> beta1v1

File: README.md [hunk 9]
```diff
@@ -1,66 +1,145 @@
...
+* `--leader-election`: Enables leader election. This is useful when there are multiple replicas of the same snapshot controller running for the same Kubernetes deployment. Only one of them may be active (=leader). A new leader will be re-elected when current leader dies or becomes unresponsive for ~15 seconds.
+
+* `--leader-election-namespace <namespace>`: The namespace where the leader election resource exists. Defaults to the pod namespace if not set.
+
+#### Other recognized arguments
+* `--kubeconfig <path>`: Path to Kubernetes client configuration that the snapshot controller uses to connect to Kubernetes API server. When omitted, default token provided by Kubernetes will be used. This option is useful only when the snapshot controller does not run as a Kubernetes pod, e.g. for debugging.
+
+* `--resync-period <duration>`: Internal resync interval when the snapshot controller re-evaluates all existing `VolumeSnapshot` instances and tries to fulfill them, i.e. create / delete corresponding snapshots. It does not affect re-tries of failed calls! It should be used only when there is a bug in Kubernetes watch logic.
```
Comment (msau42) [inline README.md]: is there a default?

File: README.md [hunk 10]
```diff
@@ -1,66 +1,145 @@
...
+
+* `--create-snapshotcontent-retrycount`: Number of retries when we create a snapshot content object for a snapshot. Default is 5.
```
Comment (msau42) [inline README.md]: does that mean we give up entirely after 5? I thought we changed it to retry forever until user cancels the operation

File: README.md (hunk 8 above)
Comment (msau42) [reply README.md]: should this go into changelog instead?

File: README.md [hunk 11]
```diff
@@ -1,66 +1,145 @@
...
+
+* `--leader-election`: Enables leader election. This is useful when there are multiple replicas of the same external-snapshotter running for one CSI driver. Only one of them may be active (=leader). A new leader will be re-elected when current leader dies or becomes unresponsive for ~15 seconds.
+
+* `--leader-election-namespace <namespace>`: The namespace where the leader election resource exists. Defaults to the pod namespace if not set.
+
+* `--timeout <duration>`: Timeout of all calls to CSI driver. It should be set to value that accommodates majority of `CreateSnapshot`, `DeleteSnapshot, and `ListSnapshots` calls. 1 minute is used by default.
+
+* `snapshot-name-prefix`: Prefix to apply to the name of a created snapshot. Default is `snapshot`.
+
+* `snapshot-name-uuid-length`: Length in characters for the generated uuid of a created snapshot. Defaults behavior is to NOT truncate.
```
Comment (msau42) [inline README.md]: Do we make sure the length doesn't exceed the CSI spec max? especially when prefix is added?

File: README.md [hunk 12]
```diff
@@ -1,66 +1,145 @@
...
+### CSI external snapshotter sidecar command line options
+
+#### Important optional arguments that are highly recommended to be used
+* `--csi-address <path to CSI socket>`: This is the path to the CSI driver socket inside the pod that the external-snapshotter container will use to issue CSI operations (`/run/csi/socket` is used by default).
+
+* `--leader-election`: Enables leader election. This is useful when there are multiple replicas of the same external-snapshotter running for one CSI driver. Only one of them may be active (=leader). A new leader will be re-elected when current leader dies or becomes unresponsive for ~15 seconds.
+
+* `--leader-election-namespace <namespace>`: The namespace where the leader election resource exists. Defaults to the pod namespace if not set.
+
+* `--timeout <duration>`: Timeout of all calls to CSI driver. It should be set to value that accommodates majority of `CreateSnapshot`, `DeleteSnapshot, and `ListSnapshots` calls. 1 minute is used by default.
```
Comment (msau42) [inline README.md]: closing quote missing after DeleteSnapshot

File: README.md [hunk 13]
```diff
@@ -1,66 +1,145 @@
...
+
+* All glog / klog arguments are supported, such as `-v <log level>` or `-alsologtostderr`.
+
+
+## Upgrade from Alpha to Beta
+
+The change from Alpha to Beta snapshot APIs is not backward compatible.
+
+If you have already deployed Alpha snapshot APIs and controller and want to upgrade to Beta, you need to do the following:
+* Delete snapshots created using Alpha snapshot CRDs and controller.
```
Comment (msau42) [inline README.md]: what does "controller" refer to in the alpha sense?

File: README.md [hunk 14]
```diff
@@ -1,66 +1,145 @@
...
---leader-election=true
-```
 
 ## Testing
 
 Running Unit Tests:
 
 ```bash
-go test -timeout 30s  github.com/kubernetes-csi/external-snapshotter/pkg/controller
+go test -timeout 30s  github.com/kubernetes-csi/external-snapshotter/pkg/common-controller
```
Comment (msau42) [inline README.md]: does "make test" accomplish the same?

File: README.md (hunk 10 above)
Comment (xing-yang) [reply README.md]: This retry count is used for creating VolumeSnapshotContent API objects.  We didn't talk about retrying this one forever.

The logic we are trying to change is for creating a physical snapshot on the storage system itself.  Grant started a PR which followed what Jan has done for the external-provisioner.  Due to the beta splitting controller work, that work is delayed.  That involves big controller logic changes which I do not want to add on top of the splitting controller changes in this release.  We can revisit that after we cut this release.

Comment (msau42) [reply README.md]: If API server is temporarily down, then we will give up on trying to bind or create a snapshot? I'm not sure we want that? Also keep in mind especially now that this is beta, we need to support the argument according to the deprecation policy.

Comment (xing-yang) [reply README.md]: I think we copied it from the PV controller. Looks like PV controller retries creating PV object 5 times by default, but createProvisionedPVRetryCount is a const defined inside the PV controller.  Should we get rid of these command line options and use a const in the snapshot controller instead?

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L131
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L1489

Comment (msau42) [reply README.md]: Yeah I think that would be better. It gives more flexibility to change the behavior later.

I think PV controller scenario is slightly different than snapshot. For snapshots, I think we have stronger idempotency/naming guarantees, so it can retry the calls with the same id, which helps (but not completely) avoids the leaking issue. So I'm not sure how necessary the retries in snapshot controller should be. cc @jsafrane

Comment (msau42) [reply README.md]: Actually, does the scenario apply at all? Now we create the Content object before calling CSI. This avoids the leaking scenario right?

Comment (xing-yang) [reply README.md]: Yes, you are right.

File: README.md (hunk 4 above)
Comment (xing-yang) [reply README.md]: Sure.  I can add that.  I actually have that in my first draft of the README but removed them when I submitted the PR because I thought maybe the workflow that @yuxiangqian is working on can better address this.

File: README.md (hunk 13 above)
Comment (xing-yang) [reply README.md]: "controller" here refers to the alpha version external-snapshotter sidecar controller.

File: README.md (hunk 8 above)
Comment (xing-yang) [reply README.md]: Sure, I'll add this to the changelog, but I still want to highlight some of these in this doc.  Let me rephrase a little bit.

File: README.md (hunk 7 above)
Comment (xing-yang) [reply README.md]: sure

File: README.md (hunk 9 above)
Comment (xing-yang) [reply README.md]: Yes, 60 sec

File: README.md (hunk 10 above)
Comment (xing-yang) [reply README.md]: Removed these from README.  Will submit a PR that removes them from cmd line options and uses a const internally instead.

File: README.md (hunk 12 above)
Comment (xing-yang) [reply README.md]: Fixed

File: README.md (hunk 11 above)
Comment (xing-yang) [reply README.md]: The CSI spec says length of the name conforms to the length limit of an unicode string.  The code we have to handle the snapshot name in the snapshotter is similar to how the volume name is handled in the provisioner.

https://github.com/kubernetes-csi/external-provisioner/blob/master/pkg/controller/controller.go#L319
https://github.com/kubernetes-csi/external-provisioner/blob/master/README.md#other-recognized-arguments

File: README.md [hunk 15]
```diff
@@ -1,66 +1,150 @@
...
+## Upgrade from v1alpha1 to v1beta1
+
+The change from v1alpha1 to v1beta1 snapshot APIs is not backward compatible.
+
+If you have already deployed v1alpha1 snapshot APIs and external-snapshotter sidecar controller and want to upgrade to v1beta1, you need to do the following:
+* Delete volume snapshots created using v1alpha1 snapshot CRDs and external-snapshotter sidecar controller.
+* Uninstall v1alpha1 snapshot CRDs, external-snapshotter sidecar controller, and CSI driver.
+* Install v1beta1 snapshot CRDs, snapshot controller, CSI external-snapshotter sidecar and CSI driver.
+
+If you want to keep the physical snapshot resources created using the v1alpha1 version, here are suggested steps.
```
Comment (xing-yang) [inline README.md]: @msau42 @yuxiangqian please review this section to make sure it is okay.

File: README.md (hunk 14 above)
Comment (xing-yang) [reply README.md]: "make test" will do more than just running unit tests.  It will also do go mod, go fmt, etc.

File: README.md (hunk 15 above)
Comment (xing-yang) [reply README.md]: @bswartz Here is the upgrade section.  PTAL.

File: README.md (hunk 11 above)
Comment (msau42) [reply README.md]: CSI Spec says 128 max: https://github.com/container-storage-interface/spec/blob/master/spec.md#size-limits.

Let's followup on this as a separate issue?

File: README.md [hunk 16]
```diff
@@ -1,66 +1,150 @@
...
+Note: There is a possibility for data loss if something goes wrong during this process. Verify steps 1 and 2 works for a sample snapshot before moving forward. There was a bug in the pre-provisioned snapshots logic that was fixed in release [v1.2.1](https://github.com/kubernetes-csi/external-snapshotter/blob/release-1.2/CHANGELOG-1.2.md#notable-changes-1). Make sure you are using v1.2.1 or later before upgrading to v1beta1 if you want to try the following steps.
+1. Change deletionPolicy from Delete to Retain in the volume snapshot content.
+2. Delete volume snapshots created using v1alpha1 snapshot CRDs and external-snapshotter sidecar controller. Verify that volume snapshot content and the physical snapshot resource remain.
+3. Make sure above steps work for a sample snapshot before moving forward. This is to make sure that deletion of a volume snapshot object will not trigger the deletion of the volume snapshot content object and its associated physical snapshot resource on the storage system when the deletionPolicy is Retain. It is very important to verify this step to make sure there is no data loss.
```
Comment (msau42) [inline README.md]: Maybe put this note as 1), since you want them to do this before they actually try the steps

File: README.md [hunk 17]
```diff
@@ -1,66 +1,150 @@
...
+4. Uninstall v1alpha1 snapshot CRDs, external-snapshotter sidecar controller, and CSI driver. The v1alpha1 volume snapshot content and its associated physical snapshot resource will not be in the Kubernetes database any more.
+5. Install v1beta1 snapshot CRDs, snapshot controller, CSI external-snapshotter sidecar, and CSI drvier.
+6. Create a v1beta1 volume snapshot content API object for the physical snapshot resource.
+  * Note: Do not set the UID field of VolumeSnapshotRef! If you copied a volume snapshot content yaml file from your v1alpha1 setup, make sure to DELETE the UID field of VolumeSnapshotRef. This field will be populated by the controller automatically after a v1beta1 volume snapshot API object is created and bound to the volume snapshot content.
```
Comment (msau42) [inline README.md]: are there any other things like annotations that they should also remove?

**Comment by xing-yang** [comment]
> CSI Spec says 128 max: https://github.com/container-storage-interface/spec/blob/master/spec.md#size-limits.

> Let's followup on this as a separate issue?

Sure.  I'll open an issue to track.

File: README.md (hunk 17 above)
Comment (xing-yang) [reply README.md]: Will remove this section as I have not tested it myself.

File: README.md (hunk 11 above)
Comment (xing-yang) [reply README.md]: Sure

Comment (xing-yang) [reply README.md]: Opened an issue here: https://github.com/kubernetes-csi/external-snapshotter/issues/216
